### ISSUE TITLE
Build server failure

### ISSUE THREAD

**Issue by jonathanolson** [phetsims/perennial#14]
Was doing RC deployments for maintenance releases (fairly rapid-fire) and saw an email alert:

```
***** Nagios *****

Notification Type: PROBLEM

Service: nodejs build-server process
Host: figaro.colorado.edu:
Address: figaro.colorado.edu
State: CRITICAL

Date/Time: Sat Mar 5 02:25:22 MST 2016

Additional Info:

PROCS CRITICAL: 0 processes with args /usr/local/nodejs/bin/node /data/share/phet/phet-repos/perennial/js/build-server/build-server.js
```

copied build-server log out, restarted build-server (via /etc/init.d/build-server start).

Notable lines:

```

/data/share/phet/phet-repos/perennial/node_modules/async/lib/async.js:30
            if (called) throw new Error("Callback was already called.");
                              ^
Error: Callback was already called.
    at /data/share/phet/phet-repos/perennial/node_modules/async/lib/async.js:30:31
    at /data/share/phet/phet-repos/perennial/js/build-server/build-server.js:308:13
    at /data/share/phet/phet-repos/perennial/js/build-server/build-server.js:295:27
    at ChildProcess.exithandler (child_process.js:646:7)
    at ChildProcess.emit (events.js:98:17)
    at maybeClose (child_process.js:756:16)
    at Socket.<anonymous> (child_process.js:969:11)
    at Socket.emit (events.js:95:17)
    at Pipe.close (net.js:465:12)
```

Looks like a potential async-related bug where our taskCallback is getting called twice.

Additionally, the fact that I had to semi-manually restart build-server worries me. We may want it to auto-restart if it dies.

@jbphet, can we discuss this on Monday?

**Comment by jonathanolson** [comment]
Didn't occur during multiple deployments today. Lowering priority.

**Comment by jbphet** [comment]
I am going to contact OIT and find out if this process can automatically be restarted if it goes down.

**Comment by jbphet** [comment]
I have contacted OIT about auto restarting, issue ID is INC0241596.

**Comment by jonathanolson** [comment]
Occurred again, just after this error:
```
Apr 01 2016 15:59:05 MDT - error: Build failed with error: error running command scp resistance-in-a-wire_nl.html jblanco@spot.colorado.edu:/htdocs/physics/phet/dev/html/resistance-in-a-wire/1.2.2-rc.1: Error: Command failed: ssh_exchange_identification: Connection closed by remote host
```

Then failed with:
```
/data/share/phet/phet-repos/perennial/node_modules/async/lib/async.js:30
            if (called) throw new Error("Callback was already called.");
                              ^
Error: Callback was already called.
    at /data/share/phet/phet-repos/perennial/node_modules/async/lib/async.js:30:31
    at /data/share/phet/phet-repos/perennial/js/build-server/build-server.js:310:13
    at /data/share/phet/phet-repos/perennial/js/build-server/build-server.js:297:27
    at ChildProcess.exithandler (child_process.js:646:7)
    at ChildProcess.emit (events.js:98:17)
    at maybeClose (child_process.js:756:16)
    at Process.ChildProcess._handle.onexit (child_process.js:823:5)
```

That might indicate where the extra callback is.

Also build-server didn't restart itself.

**Comment by jbphet** [comment]
Build server should now be set up to be automatically restarted if it goes down.

**Comment by jbphet** [comment]
The build server has now been migrated to our new server machine (phet-server.colorado.edu), and we are using systemctl to handle task management.  It's been stable there for a while, so I think it's safe to close this.

**Comment by zepumph** [comment]
This just came up today while trying to deploy a new RC version with @andrea-phet for projectile-motion. We tried to deploy 4 separate times:

1. AL ran `grunt deploy-rc` but with an incorrect authorization code, so we got the following error in the build-server logs: `error: wrong authorization code`.

2. We fixed the authorization code in build-local.json and tried again, here is the end of the log:
```bash
Aug 15 17:51:24 phet-server.int.colorado.edu build-server[61194]: Aug 15 2017 17:51:24 MDT - info: ssh jblanco@spot.colorado.edu 'chmod -R g+w /htdocs/physics/phet/dev/html/projectilAug 15 17:51:24 phet-server.int.colorado.edu build-server[61194]: Aug 15 2017 17:51:24 MDT - info: running command: grunt checkout-master-all
Aug 15 17:51:24 phet-server.int.colorado.edu build-server[61194]: Aug 15 2017 17:51:24 MDT - info: grunt checkout-master-all ran successfully in directory: .
Aug 15 17:51:24 phet-server.int.colorado.edu build-server[61194]: Aug 15 2017 17:51:24 MDT - info: running command: rm -rf ./js/build-server/tmp
Aug 15 17:51:24 phet-server.int.colorado.edu build-server[61194]: Aug 15 2017 17:51:24 MDT - info: rm -rf ./js/build-server/tmp ran successfully in directory: .
Aug 15 17:51:24 phet-server.int.colorado.edu build-server[61194]: Aug 15 2017 17:51:24 MDT - info: build for projectile-motion finished successfully
Aug 15 17:51:24 phet-server.int.colorado.edu build-server[61194]: Aug 15 2017 17:51:24 MDT - info: attempting to send email
Aug 15 17:51:24 phet-server.int.colorado.edu build-server[61194]: Aug 15 2017 17:51:24 MDT - info: sent email to: andrealin2018@gmail.com, subject: Build Succeeded, text: Sim = projeAug 15 17:51:25 phet-server.int.colorado.edu build-server[61194]: Aug 15 2017 17:51:25 MDT - info: grunt checkout-master-all ran successfully in directory: .
Aug 15 17:51:25 phet-server.int.colorado.edu build-server[61194]: Aug 15 2017 17:51:25 MDT - info: running command: rm -rf ./js/build-server/tmp
Aug 15 17:51:25 phet-server.int.colorado.edu build-server[61194]: Aug 15 2017 17:51:25 MDT - info: rm -rf ./js/build-server/tmp ran successfully in directory: .
Aug 15 17:51:25 phet-server.int.colorado.edu build-server[61194]: /data/share/phet/phet-repos/perennial/node_modules/async/lib/async.js:30
Aug 15 17:51:25 phet-server.int.colorado.edu build-server[61194]: if (called) throw new Error("Callback was already called.");
Aug 15 17:51:25 phet-server.int.colorado.edu build-server[61194]: ^
Aug 15 17:51:25 phet-server.int.colorado.edu build-server[61194]: Error: Callback was already called.
Aug 15 17:51:25 phet-server.int.colorado.edu build-server[61194]: at /data/share/phet/phet-repos/perennial/node_modules/async/lib/async.js:30:31
Aug 15 17:51:25 phet-server.int.colorado.edu build-server[61194]: at /data/share/phet/phet-repos/perennial/js/build-server/build-server.js:869:9
Aug 15 17:51:25 phet-server.int.colorado.edu build-server[61194]: at /data/share/phet/phet-repos/perennial/js/build-server/build-server.js:335:27
Aug 15 17:51:25 phet-server.int.colorado.edu build-server[61194]: at ChildProcess.exithandler (child_process.js:204:7)
Aug 15 17:51:25 phet-server.int.colorado.edu build-server[61194]: at emitTwo (events.js:87:13)
Aug 15 17:51:25 phet-server.int.colorado.edu build-server[61194]: at ChildProcess.emit (events.js:172:7)
Aug 15 17:51:25 phet-server.int.colorado.edu build-server[61194]: at maybeClose (internal/child_process.js:829:16)
Aug 15 17:51:25 phet-server.int.colorado.edu build-server[61194]: at Socket.<anonymous> (internal/child_process.js:319:11)
Aug 15 17:51:25 phet-server.int.colorado.edu build-server[61194]: at emitOne (events.js:77:13)
Aug 15 17:51:25 phet-server.int.colorado.edu build-server[61194]: at Socket.emit (events.js:169:7)
Aug 15 17:51:25 phet-server.int.colorado.edu systemd[1]: build-server.service: main process exited, code=exited, status=1/FAILURE
Aug 15 17:51:25 phet-server.int.colorado.edu systemd[1]: Unit build-server.service entered failed state.
Aug 15 17:51:25 phet-server.int.colorado.edu systemd[1]: build-server.service failed.
Aug 15 17:51:25 phet-server.int.colorado.edu systemd[1]: build-server.service holdoff time over, scheduling restart.
Aug 15 17:51:25 phet-server.int.colorado.edu systemd[1]: Started build-server.service.
Aug 15 17:51:25 phet-server.int.colorado.edu systemd[1]: Starting build-server.service...
```

3. Tried again and got the same error.

4. I tried it on my side after pulling her commits to the new branch and what not. We got the same error, but then I got an email saying success and the rc showed up on phet.

```bash
Aug 15 18:04:08 phet-server.int.colorado.edu build-server[143904]: Aug 15 2017 18:04:08 MDT - info: running command: grunt checkout-master-all
Aug 15 18:04:08 phet-server.int.colorado.edu build-server[143904]: Aug 15 2017 18:04:08 MDT - info: ssh jblanco@spot.colorado.edu 'chmod -R g+w /htdocs/physics/phet/dev/html/projectiAug 15 18:04:08 phet-server.int.colorado.edu build-server[143904]: Aug 15 2017 18:04:08 MDT - info: running command: grunt checkout-master-all
Aug 15 18:04:09 phet-server.int.colorado.edu build-server[143904]: Aug 15 2017 18:04:09 MDT - info: grunt checkout-master-all ran successfully in directory: .
Aug 15 18:04:09 phet-server.int.colorado.edu build-server[143904]: Aug 15 2017 18:04:09 MDT - info: running command: rm -rf ./js/build-server/tmp
Aug 15 18:04:09 phet-server.int.colorado.edu build-server[143904]: Aug 15 2017 18:04:09 MDT - info: rm -rf ./js/build-server/tmp ran successfully in directory: .
Aug 15 18:04:09 phet-server.int.colorado.edu build-server[143904]: Aug 15 2017 18:04:09 MDT - info: build for projectile-motion finished successfully
Aug 15 18:04:09 phet-server.int.colorado.edu build-server[143904]: Aug 15 2017 18:04:09 MDT - info: attempting to send email
Aug 15 18:04:09 phet-server.int.colorado.edu build-server[143904]: Aug 15 2017 18:04:09 MDT - info: sent email to: michael.kauzmann@colorado.edu, subject: Build Succeeded, text: Sim Aug 15 18:04:09 phet-server.int.colorado.edu build-server[143904]: Aug 15 2017 18:04:09 MDT - info: grunt checkout-master-all ran successfully in directory: .
Aug 15 18:04:09 phet-server.int.colorado.edu build-server[143904]: Aug 15 2017 18:04:09 MDT - info: running command: rm -rf ./js/build-server/tmp
Aug 15 18:04:09 phet-server.int.colorado.edu build-server[143904]: Aug 15 2017 18:04:09 MDT - info: rm -rf ./js/build-server/tmp ran successfully in directory: .
Aug 15 18:04:09 phet-server.int.colorado.edu build-server[143904]: /data/share/phet/phet-repos/perennial/node_modules/async/lib/async.js:30
Aug 15 18:04:09 phet-server.int.colorado.edu build-server[143904]: if (called) throw new Error("Callback was already called.");
Aug 15 18:04:09 phet-server.int.colorado.edu build-server[143904]: ^
Aug 15 18:04:09 phet-server.int.colorado.edu build-server[143904]: Error: Callback was already called.
Aug 15 18:04:09 phet-server.int.colorado.edu build-server[143904]: at /data/share/phet/phet-repos/perennial/node_modules/async/lib/async.js:30:31
Aug 15 18:04:09 phet-server.int.colorado.edu build-server[143904]: at /data/share/phet/phet-repos/perennial/js/build-server/build-server.js:869:9
Aug 15 18:04:09 phet-server.int.colorado.edu build-server[143904]: at /data/share/phet/phet-repos/perennial/js/build-server/build-server.js:335:27
Aug 15 18:04:09 phet-server.int.colorado.edu build-server[143904]: at ChildProcess.exithandler (child_process.js:204:7)
Aug 15 18:04:09 phet-server.int.colorado.edu build-server[143904]: at emitTwo (events.js:87:13)
Aug 15 18:04:09 phet-server.int.colorado.edu build-server[143904]: at ChildProcess.emit (events.js:172:7)
Aug 15 18:04:09 phet-server.int.colorado.edu build-server[143904]: at maybeClose (internal/child_process.js:829:16)
Aug 15 18:04:09 phet-server.int.colorado.edu build-server[143904]: at Process.ChildProcess._handle.onexit (internal/child_process.js:211:5)
Aug 15 18:04:09 phet-server.int.colorado.edu systemd[1]: build-server.service: main process exited, code=exited, status=1/FAILURE
Aug 15 18:04:09 phet-server.int.colorado.edu systemd[1]: Unit build-server.service entered failed state.
Aug 15 18:04:09 phet-server.int.colorado.edu systemd[1]: build-server.service failed.
Aug 15 18:04:09 phet-server.int.colorado.edu systemd[1]: build-server.service holdoff time over, scheduling restart.
Aug 15 18:04:09 phet-server.int.colorado.edu systemd[1]: Started build-server.service.
Aug 15 18:04:09 phet-server.int.colorado.edu systemd[1]: Starting build-server.service...
Aug 15 18:04:09 phet-server.int.colorado.edu build-server[147058]: Aug 15 2017 18:04:09 MDT - info: Listening on port 16371
Aug 15 18:04:09 phet-server.int.colorado.edu build-server[147058]: Aug 15 2017 18:04:09 MDT - info: Verbose mode: false

```

This makes me think that it is possible that we have gotten this error in the past, but no one knew because the build succeeded and the service restarted itself.

**Comment by zepumph** [comment]
I ran `sudo journalctl -u  build-server | grep "Error: Callback was already called."` and found 173 lines since January that matched. It looks like this is an ongoing problem.

**Comment by jbphet** [comment]
I just logged into phet-server and ran the command @zepumph described for the previous test and found no instance of the error string.  I examined the log and saw the following line in it, which indicates the time span:

```
-- Logs begin at Tue 2019-03-12 01:04:01 MDT, end at Tue 2020-01-07 13:24:05 MST. --
```

Based on this, I think it's safe to say that the problem hasn't occurred in over nine months, and no one has updated this issue for over two years, or run into the problem that I'm aware of.  I'm going to go ahead and close it.
